airbyte-integrations/connectors/source-railz/source_railz/components.py

```diff
@@ -0,0 +1,140 @@
+from airbyte_cdk.sources.declarative.types import Config, Record, StreamSlice
+from airbyte_cdk.sources.streams.http.requests_native_auth.abstract_token import AbstractHeaderAuthenticator
+from dataclasses_jsonschema import JsonSchemaMixin
+from isodate import Duration, parse_duration
+
+
+@dataclass
+class ShortLivedTokenAuthenticator(AbstractHeaderAuthenticator, DeclarativeAuthenticator, JsonSchemaMixin):
+    """
+    https://docs.railz.ai/reference/authentication
+    """
+
+    client_id: Union[InterpolatedString, str]
+    secret_key: Union[InterpolatedString, str]
+    url: Union[InterpolatedString, str]
+    config: Config
+    options: InitVar[Mapping[str, Any]]
+    token_key: Union[InterpolatedString, str] = "access_token"
+    lifetime: Union[InterpolatedString, str] = "PT3600S"
+
+    def __post_init__(self, options: Mapping[str, Any]):
+        self._client_id = InterpolatedString.create(self.client_id, options=options)
+        self._secret_key = InterpolatedString.create(self.secret_key, options=options)
+        self._url = InterpolatedString.create(self.url, options=options)
+        self._token_key = InterpolatedString.create(self.token_key, options=options)
+        self._lifetime = InterpolatedString.create(self.lifetime, options=options)
+        self._basic_auth = BasicHttpAuthenticator(
+            username=self._client_id,
+            password=self._secret_key,
+            config=self.config,
+            options=options,
+        )
+        self._session = requests.Session()
+        self._token = None
+        self._timestamp = None
+
+    @classmethod
+    def _parse_timedelta(cls, time_str) -> Union[datetime.timedelta, Duration]:
+        """
+        :return Parses an ISO 8601 durations into datetime.timedelta or Duration objects.
+        """
+        if not time_str:
+            return datetime.timedelta(0)
+        return parse_duration(time_str)
+
+    def check_token(self):
+        now = time.time()
+        url = self._url.eval(self.config)
+        token_key = self._token_key.eval(self.config)
+        lifetime = self._parse_timedelta(self._lifetime.eval(self.config))
+        if not self._token or now - self._timestamp >= lifetime.seconds:
+            response = self._session.get(url, headers=self._basic_auth.get_auth_header())
+            response.raise_for_status()
+            response_json = response.json()
+            if token_key not in response_json:
+                raise Exception(f"token_key: '{token_key}' not found in response {url}")
+            self._token = response_json[token_key]
+            self._timestamp = now
+
+    @property
+    def auth_header(self) -> str:
+        return "Authorization"
+
+    @property
+    def token(self) -> str:
+        self.check_token()
+        return f"Bearer {self._token}"
+
+
+@dataclass
+class RailzCartesianProductStreamSlicer(CartesianProductStreamSlicer):
+    """
+    Some streams require support of nested state:
```

Comment:
why does the state need to be nested as opposed to a flat state?
```
{
"accounting_transactions": {
  "business_name": "Business1",
  "service_name": "dynamicsBusinessCentral",
  "postedDate": "2022-12-28T00:00:00.000Z"
  }
}
```

Reply:
In stream I have multiple pairs `business_name` and `service_name` and  for every pair, I need to keep separate `postedDate`

Like this one https://github.com/airbytehq/airbyte/blob/f4c2fcd652266fa6611a155d5b200925f8fddb0d/airbyte-integrations/connectors/source-railz/integration_tests/abnormal_state.json#L9

We have something similar in `source-github` connector:
https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-github/integration_tests/abnormal_state.json#L16

Reply:
@girarda ping

Reply:
@brianjlai is this going to be a viable pattern after splitting incremental and iteration?